### DISCUSSION TITLE
MongoDB "OneToOne" Not Working

### DISCUSSION THREAD

**Discussion by necm1** [mikro-orm/mikro-orm#2498] [Q&A]
**Describe the bug**
Tried multiple ways to get a relationship working, but it looks like it seems not to be working for Mongo

**To Reproduce**
1. Create following tables / documents / collections:

User:
```
    "_id": {
        "$oid": "ID"
    },
    "username": "necm1",
```

Profile
```
{
    "_id": {
        "$oid": "ID"
    },
    "user_id": {
        "$oid": "USER_ID"
    },
}
```
Defining `user_id` as `string` and setting the ID from the `users` table / document / collection seems also not to work

2. Create Entities

User
```
@Entity({collection: 'users'})
export class User {
  @PrimaryKey()
  _id: ObjectId;

  @OneToOne({
    entity: () => Profile,
    owner: true,
    orphanRemoval: true,
    cascade: [],
    eager: true,
    mappedBy: (profile) => profile.user,
  })
  public readonly profile: Profile;

    // OR
    @OneToOne(() => Profile, (profile) => profile.user, {
     owner: true,
     orphanRemoval: true,
    })
   profile!: Profile;
}
```

Profile:
```
@Entity({collection: 'users_profiles'})
export class Profile {
  @PrimaryKey()
  _id: ObjectId;

  @OneToOne({entity: () => User})
  user: User;

  // OR

  @OneToOne(() => User, (user) => user.profile)
  user: User;
}
```

3. Logging `user.profile` results to `undefined`

Code Example (NestJS):

```
  /**
   * UserService constructor
   *
   * @constructor
   * @param userRepository
   */
  constructor(
    @InjectRepository(User)
    private readonly userRepository: EntityRepository<User>
  ) {}

  /**
   * Get user by name
   *
   * @public
   * @param username
   * @returns Promise<User>
   */
  public async getByName(username: string): Promise<User> {
    // even this isn't working: await this.userRepository.findOne({username}, ['profile']);
    const user = await this.userRepository.findOne({username});

    // its undefined
    console.log(user.profile);

    if (!user) {
      throw new UserNotFoundException();
    }

    return user;
  }

```

**Versions**
```
    "@mikro-orm/core": "^4.5.9",
    "@mikro-orm/mongodb": "^4.5.9",
    "@mikro-orm/nestjs": "^4.3.0",

```

**Comment by B4nan** [comment]
I have no idea what you mean by the $oid thing in your data example. Also you say something is not working, but you havent share any actual code that would work with the entities, nor an error example you see. 

Where do you want that foreign key to be stored? 

Setting cascade to empty array (so no cascades) is literally the opposite to using orphan removal (as that implies cascade.all). 

Also note that you are supposed to say what exact versions are you using, saying just latest wont help much, latest version will keep changing.

**Reply by necm1** [reply]
> I have no idea what you mean by the $oid thing in your data example

`$oid` stands for `ObjectID` / `Object Identifier`

> Also you say something is not working, but you havent share any actual code that would work with the entities, nor an error example you see.

I clearly wrote it in the title & kinda in the post. Could be my mistake, so sorry for that.

First post:
> Tried multiple ways to get a relationship working, 

Title:
> MongoDB "OneToOne" Not Working

If it's still not clear enough; I'm trying to use `OneToOne` with MongoDB and somehow I cannot make it work.

---
>  but you havent share any actual code that would work with the entities, nor an error example you see.

I mentioned that I'm getting `undefined` as the result. Code-Example using NestJS:
```
  /**
   * UserService constructor
   *
   * @constructor
   * @param userRepository
   */
  constructor(
    @InjectRepository(User)
    private readonly userRepository: EntityRepository<User>
  ) {}

  /**
   * Get user by name
   *
   * @public
   * @param username
   * @returns Promise<User>
   */
  public async getByName(username: string): Promise<User> {
    // even this isn't working: await this.userRepository.findOne({username}, ['profile']);
    const user = await this.userRepository.findOne({username});

    // its undefined
    console.log(user.profile);

    if (!user) {
      throw new UserNotFoundException();
    }

    return user;
  }
```

> Setting cascade to empty array (so no cascades) is literally the opposite to using orphan removal (as that implies cascade.all).

Actually it was only to test if `OneToOne` works, but thanks for the info.

> Also note that you are supposed to say what exact versions are you using, saying just latest wont help much, latest version will keep changing.

That's true and I'm sorry for that. Gonna edit my first post too!

Current versions:
```
    "@mikro-orm/core": "^4.5.9",
    "@mikro-orm/mongodb": "^4.5.9",
    "@mikro-orm/nestjs": "^4.3.0",
```

**Comment by necm1** [comment]
I've found a really strange way to load the `OneToOne` relationship. Either I'm doing it wrong or it's a bug.

profile.entity.ts (not owner)
```
export class Profile {
  @PrimaryKey()
  _id: ObjectId;

  @OneToOne({entity: () => User, mappedBy: 'profile', fieldName: 'user_id'})
  user: User;
}
```

user.entity.ts (owner)
```
export class User {
  @PrimaryKey()
  _id: ObjectId;

 @Property()
 username: string;

  @OneToOne({
    entity: () => Profile,
    owner: true,
    eager: true,
  })
  profile: Profile;
}
```

Loading it how it should be loaded (I guess)
```
export class UserService {
  /**
   * UserService constructor
   *
   * @constructor
   * @param userRepository
   */
  constructor(
    @InjectRepository(User)
    private readonly userRepository: EntityRepository<User>,
    @InjectRepository(Profile)
    private readonly profile: EntityRepository<Profile>
  ) {}


  /**
   * Get user by name
   *
   * @public
   * @param username
   * @returns Promise<User>
   */
  public async getByName(username: string): Promise<User> {
    // Especially telling to load the relationship while using eager
    // so I can make sure its 100% included
    const user = await this.userRepository.findOne({username: 'test'}, ['profile']);

    // I need to add this otherwise profile is never loaded
    await this.profile.findAll();

    // print output to console
    console.log(user);

    if (!user) {
      throw new UserNotFoundException();
    }

    return user;
  }
}

```

Console Output:
```
[query] db.getCollection('users').find({ username: 'test' }, { session: undefined }).limit(1).toArray(); [took 4 ms]
[query] db.getCollection('users_profiles').find({}, { session: undefined }).toArray(); [took 1 ms]
User {
  _id: ObjectId('61ab8c2695416da67e475359'),
  username: 'test',
  profile: Profile {
    _id: ObjectId('61ab9a0995416da67e475362'),
    user: User {
      _id: ObjectId('61ab8c2695416da67e475359'),
      username: 'test',
      profile: [Profile]
    },
  }
}
```

but if we comment `await this.profile.findAll()` out, it prints this out:
```
[query] db.getCollection('users').find({ username: 'test' }, { session: undefined }).limit(1).toArray(); [took 4 ms]
User {
  _id: ObjectId('61ab8c2695416da67e475359'),
  username: 'test',
}
```

Why is `profile` gone now? Does the `OneToOne`-Relation fail to load the data before? Or does it require that the data is loaded before? If so, then it wouldn't make much sense.

It doesn't help to add `mappedBy: 'user'` to the owner (User) like this:
```
  @OneToOne({
    entity: () => Profile,
    mappedBy: 'user',
    owner: true,
    eager: true,
  })
  profile: Profile;
```

**Reply by B4nan** [reply]
Well, it works just fine for me, here is a passing test case:

```ts
import { ObjectId } from 'mongodb';
import { Entity, PrimaryKey, Property, MikroORM, OneToOne } from '@mikro-orm/core';

@Entity()
class User {

  @PrimaryKey()
  _id!: ObjectId;

  @Property()
  username!: string;

  @OneToOne({
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    entity: () => Profile,
    owner: true,
    eager: true,
  })
  profile!: any; // can't use Profile here due to cycles, would work if entities are in separate files

}

@Entity()
class Profile {

  @PrimaryKey()
  _id!: ObjectId;

  @OneToOne({ entity: () => User, mappedBy: 'profile' })
  user?: User;

}

describe('GH issue 2498', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [User, Profile],
      clientUrl: 'mongodb://localhost:27017,localhost:27018,localhost:27019/mikro-orm-test?replicaSet=rs0',
      type: 'mongo',
      debug: true,
    });
    await orm.em.nativeDelete(User, {});
    await orm.em.nativeDelete(Profile, {});
  });

  afterAll(() => orm.close(true));

  test('1:1 relation loading', async () => {
    const user = new User();
    user.username = 'abc';
    user.profile = new Profile();
    await orm.em.fork().persistAndFlush(user);

    const u = await orm.em.findOneOrFail(User, user._id);
    expect(u.profile).not.toBeUndefined();
    console.log(u);
  });

});
```

This will log:

```
[query] db.getCollection('profile').insertOne({}, { session: undefined }); [took 1 ms]
[query] db.getCollection('user').insertOne({ username: 'abc', profile: ObjectId('61ac7d66ee9d0ed39359d396') }, { session: undefined }); [took 0 ms]
[query] db.getCollection('user').find({ _id: ObjectId('61ac7d66ee9d0ed39359d397') }, { session: undefined }).limit(1).toArray(); [took 1 ms]
[query] db.getCollection('profile').find({ _id: { '$in': [ ObjectId('61ac7d66ee9d0ed39359d396') ] } }, { session: undefined }).sort({ _id: 1 }).toArray(); [took 1 ms]
User {
  _id: ObjectId('61ac7d66ee9d0ed39359d397'),
  username: 'abc',
  profile: Profile {
    _id: ObjectId('61ac7d66ee9d0ed39359d396'),
    user: User {
      _id: ObjectId('61ac7d66ee9d0ed39359d397'),
      username: 'abc',
      profile: [Profile]
    }
  }
}
```

Next time if you want to report bugs, would be good if you provide such test case, that way it is 100% clear what you are doing. It does not have to be a test case per se, but it needs to be complete, so include both the initial phase where you set up the database, as well as some assertions (can be just console.logs with comments about what you expect differently).

Note that I removed the `fieldName` you have, as that does not make sense, there is no FK on inverse side, so you cant set its field name.

Also note that in mongo there are no joins, so you cant load 1:1 relation from inverse side - as there is no FK, we dont know what to load. Therefore eager: true on inverse side won't work in mongo. So what you are describing here sounds like you somehow managed to define the relation as owning on wrong side - but the entity definitions are correct, so either you are saying something else than you actually use, or you might just misobserve things, e.g. its common that you have outdated state in either metadata cache or in dist (so compiled TS entities - you can have correct TS definition but your compiled JS is still the old code).

**Reply by necm1** [reply]
> 

I don't know. I tried it at all but its not working like you described it. I copied your code 1:1 without changing anything. My `User` entity doesn't want to load `Profile` at all and vice versa. I'm using NestJS, so don't worry about the code.

AppModule (Initializing)
```
@Module({
  imports: [
    MikroOrmModule.forRoot(
       entities: ['./dist/*/entity'],
       entitiesTs: ['./src/*/entity'],
       autoLoadEntities: true,
       dbName: 'test',
      clientUrl: 'mongodb://user:test@127.0.0.1:27017/',
      type: 'mongo',
      debug: true,
    ),
    UserModule,
    CacheModule,
    AuthModule,
    ProfileModule,
  ],
  controllers: [AppController],
  providers: [],
})
/**
 * @class AppModule
 */
export class AppModule {}
```

Console output after starting NestJS:
```
[discovery] - processing 0 files
[discovery] - processing entity Profile
[discovery] - processing entity User
[discovery] - entity discovery finished, found 2 entities, took 31 ms
[info] MikroORM successfully connected to database test on mongodb://user:*****@127.0.0.1:27017/
```

Controller:
```
  /**
   * Get current signed in user
   *
   * @public
   * @async
   * @param {Request} req
   * @returns Promise<User>
   */
  public async user(@Request() req): Promise<User> {
    // loading user trough his username works fine
    return this.userService.getByName(req.user.username);
  }
```

The console output visiting `/user` is this: `[query] db.getCollection('users').find({ username: 'necm1' }, { session: undefined }).limit(1).toArray(); [took 4 ms]`

UserService (which uses `@InjectRepository`):
```
  /**
   * Get user by name
   *
   * @public
   * @param username
   * @returns Promise<User>
   */
  public async getByName(username: string): Promise<User> {
    const user = await this.userRepository.findOne({username});

    if (!user) {
      throw new UserNotFoundException();
    }

    return user;
  }
```

User Entity:
```
@Entity({collection: 'users'})
/**
 * @class User
 */
export class User {
  @PrimaryKey({hidden: true})
  _id: ObjectId;

  @OneToOne({
    entity: () => Profile,
    owner: true,
    eager: true,
  })
  profile!: any;
}
```

Profile Entity:
```
@Entity({collection: 'users_profiles'})
/**
 * @class Profile
 */
export class Profile {
  @PrimaryKey({hidden: true})
  /**
   * @public
   * @property
   * @type {ObjectId}
   */
  public readonly _id!: ObjectId;

  @OneToOne({entity: () => User, mappedBy: 'profile'})
  user?: User;
}
```

**What did I expected?**
The same output as you got here:
`[query] db.getCollection('profile').find({ _id: { '$in': [ ObjectId('61ac7d66ee9d0ed39359d396') ] } }, { session: undefined }).sort({ _id: 1 }).toArray(); [took 1 ms]` and I'm not sure how it loads the user. Does Mikro-Orm uses `entity_id` as column?

In my example it should check if `user_id` from `users_profiles` should be the same as the `_id` in `users`

**Reply by Fire3galaxy** [reply]
> Also note that in mongo there are no joins, so you cant load 1:1 relation from inverse side - as there is no FK, we dont know what to load. Therefore eager: true on inverse side won't work in mongo. So what you are describing here sounds like you somehow managed to define the relation as owning on wrong side - but the entity definitions are correct, so either you are saying something else than you actually use, or you might just misobserve things, e.g. its common that you have outdated state in either metadata cache or in dist (so compiled TS entities - you can have correct TS definition but your compiled JS is still the old code).

Hello @B4nan, I am working with MikroORM and mongodb and ran into an issue with OneOnOne decorators and the entity not loading as expected, similar to, but apparently not quite the same as @necm1. This particular issue helped me determine a possible root cause for my issue, so thank you both of you for this discussion. To clarify, my issue is that, for OneToOne relations, EntityManager::find() populates the class field for the owning side, but not for the inverse side. Based on your comment above, it sounds like that's expected.

In other words, the following code snippet not working would be expected due to this limitation.
```
test('1:1 relation loading', async () => {
    const user = new User();
    user.username = 'abc';
    user.profile = new Profile();
    await orm.em.fork().persistAndFlush(user);

    // This onward is modified for this discussion.
    const p = await orm.em.find(Profile, user.profile._id); // Only one profile right now
    expect(p).toHaveLength(1);

    // Right now, this fails for me unless the below comment is uncommented. 
    // Profile is the inverse side and thus the Mongo Entity Manager is unable to populate
    // p[0].user. On the other hand, if the below find() call is uncommented, the Entity Manager
    // does some magic and now p[0].user is populated. Is that right?
    // await orm.em.find(User, user._id);
    expect(p[0].user).not.toBeUndefined();

    console.log(u);
  });
```

I notice the issue isn't present when I use SQLite however, so to have to add an additional call for the mongo case isn't ideal. Is my interpretation correct?

**Reply by edorgeville** [reply]
Damn, @Fire3galaxy, you nailed it. I think I'm experiencing exactly this issue with MongoDB. Inverse side of `OneToOne` has an `undefined` relation unless the entity is loaded separately.

**Reply by B4nan** [reply]
Yes, this is a limitation of mongo, as if you load only the inverse side, there is no FK. If you load that owning entity separately, they will get automatically wired. With SQL drivers there is an implicit left join to get that FK to get around this, but with mongo there are no joins and `$lookup` can't be used for performance reasons. For the same reason you can't filter by that property - it does not exist in the collection.

**Reply by edorgeville** [reply]
Gotcha. Workaround would be to do a `find` manually, correct?

**Comment by Fire3galaxy** [comment]
That's how I understand it. This may be my limited experience with
databases, but I don't understand why the issue doesn't seem to apply to
one to many relationships.
